backend-tests/tests/test_deployments.py

```diff
@@ -1105,3 +1137,299 @@ def test_two_phases_concurrent_devices(self, clean_mongo):
+        )
+        resp.status_code == 200
+        devices = resp.json()
+        for device in devices:
+            if device["id"] == device_id:
+                assert device["status"] == device_deployment_status
+
+                if substate != "":
+                    assert device["substate"] == substate
+
+    if deployment_status != "":
+        resp = deploymentsm.with_auth(user_token).call(
+            "GET", deployments.URL_DEPLOYMENTS.format(id=deployment_id)
+        )
+        resp.status_code == 200
+        assert resp.json()[0]["status"] == deployment_status
+
+
+class TestDeploymentsStatusUpdateBase:
+    def do_test_deployment_status_update(self, clean_mongo, user_token, devs):
+        """
+        deployment with four devices
+        requires five devices (last one won't be part of the deployment
+        """
+
+        deploymentsm = ApiClient(deployments.URL_MGMT)
+
+        # Make deployment request
+        deployment_req = {
+            "name": "phased-deployment",
+            "artifact_name": "deployments-phase-testing",
+            "devices": [dev.id for dev in devs[:-1]],
+        }
+        resp = deploymentsm.with_auth(user_token).call(
+            "POST", deployments.URL_DEPLOYMENTS, deployment_req
+        )
+        assert resp.status_code == 201
+
+        # Store the location from the GET /deployments/{id} request
+        deployment_id = os.path.basename(resp.headers["Location"])
+
+        # Verify that the deployment is in "pending" state
+        status_verify(
+            deploymentsm, deployment_id, user_token, deployment_status="pending",
+        )
+
+        default_artifact_name = "bugs-bunny"
+        default_device_type = "qemux86-64"
+        deploymentsd = ApiClient(deployments.URL_DEVICES)
+
+        # Try to retrieve next update and assert expected status code
+        resp = deploymentsd.with_auth(devs[0].token).call(
+            "GET",
+            deployments.URL_NEXT,
+            qs_params={
+                "artifact_name": getattr(
+                    devs[0], "artifact_name", default_artifact_name
+                ),
+                "device_type": getattr(devs[0], "device_type", default_device_type),
+            },
+        )
+        assert resp.status_code == 200
+
+        # Try to retrieve next update for the device that already has the artifact
+        resp = deploymentsd.with_auth(devs[1].token).call(
+            "GET",
+            deployments.URL_NEXT,
+            qs_params={
+                "artifact_name": "deployments-phase-testing",
+                "device_type": getattr(devs[1], "device_type", default_device_type),
+            },
+        )
+        assert resp.status_code == 204
+        status_verify(
+            deploymentsm,
+            deployment_id,
+            user_token,
+            device_id=devs[1].id,
+            device_deployment_status="already-installed",
+            deployment_status="inprogress",
+        )
```

Comment:
I suggest to refactor status_verify and status_update_and_verify as methods exposed by a helper class, something among the lines of...

```
class StatusVerifier:
    def __init__(self, deploymentsm, deploymentsd):
        self.deploymentsm = deploymentsm
        self.deploymentsd = deploymentsd

    def status_verify(self, deployment_id, user_token, ...):
        ...
````

In this way we can instantiate the helper once and use it multiple times:

```
status_verifier = StatusVerifier(deploymentsm, deploymentsd)
...
status_verifier.status_verify(
    user_token,
    device_id=devs[1].id,
    device_deployment_status="already-installed",
    deployment_status="inprogress",
)
```
To make it more readable, I'd always use keyword arguments `deployment_status="inprogress"` instead of positional arguments.

Feel free to ignore this suggestion, it doesn't really matter and the test is just fine.

Reply:
@tranchitella good suggestions; thanks!
I've applied them in a separate commit - https://github.com/mendersoftware/integration/pull/816/commits/e02ab157205971c366442795d4b008302d1d8921

Reply:
Great!